### PR TITLE
Add docs for inline arrays in C# 12

### LINKED ISSUE #36066: [C# 12]: New Feature - Inline arrays
### Visual Studio release

17.7p3

### Link to speclet

- [Speclet](https://github.com/dotnet/csharplang/blob/main/proposals/inline-arrays.md)

### Link to Pull request

- [PR](https://github.com/dotnet/roslyn/pull/68783)

### New behavior

Provide a general-purpose and safe mechanism for consuming struct types utilizing https://github.com/dotnet/runtime/issues/61135 feature. Provide a general-purpose and safe mechanism for declaring inline arrays within C# classes, structs, and interfaces.

To have targets for the next preview release, split this into 2 PRs. The first for the language reference and "What's New". The second will add the error / warning messages and the tutorials.

### Areas affected

- [X] **Add to What's new in C# 12**: Add links to updated docs in the What's new in C# 12 article.
- [X] **Update language reference**: Detail the areas in the language reference that will be updated for this feature.
- [X] **Update compiler messages**: Add new (or updated) pages for new compiler messages. Include related message.
- [ ] *Include in *Everyday C#* section**: Is this new feature included in Everyday C#?
- [X] **Tutorials**: Are new tutorials required for this feature? If so, detail below

### Language reference tasks

- Struct declarations can now include "buffer like types"
- This should link to the `InlineArray` attribute
- The ref safety rules for these types should be checked.

### Tutorial tasks

- The tutorial should show how to consume these types using `Span<T>` and `ReadonlySpan<T>`.
- 

### Everyday C# tasks

This feature won't be part of the Everyday C# cohort.

---
[Associated WorkItem - 123084](https://dev.azure.com/msft-skilling/Content/_workitems/edit/123084)

### LINKED ISSUE #36134: Missing dot
There is a missing dot :)

image

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: c36588d0-6bdf-1740-220c-10b20bb88730
* Version Independent ID: 587ed47f-f532-e018-9b0d-6ee76afd20e4
* Content: [What's new in C# 12 - C# Guide](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12)
* Content Source: [docs/csharp/whats-new/csharp-12.md](https://github.com/dotnet/docs/blob/main/docs/csharp/whats-new/csharp-12.md)
* Product: **dotnet-csharp**
* Technology: **csharp-whats-new**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Contributes to #36066
Fixes #36134 

- Publish the feature speclet.
- Add new paragraphs in the What's new in C# 12 for the new feature.
- Fix grammar nit reported while editing these files.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/builtin-types/struct.md](https://github.com/dotnet/docs/blob/d8852b873f5c437b16ab69feb45e6685e216b4fa/docs/csharp/language-reference/builtin-types/struct.md) | [Structure types (C# reference)](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct?branch=pr-en-us-36114) |
| [docs/csharp/whats-new/csharp-12.md](https://github.com/dotnet/docs/blob/d8852b873f5c437b16ab69feb45e6685e216b4fa/docs/csharp/whats-new/csharp-12.md) | [What's new in C# 12 - C# Guide](https://review.learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12?branch=pr-en-us-36114) |

### CODE REVIEW

**Comment by BillWagner** [comment]
closing and reopening to test snippets build fix.

File: docs/csharp/language-reference/builtin-types/struct.md [hunk 1]
```diff
@@ -82,6 +82,28 @@ Beginning with C# 10, you can use the [`with` expression](../operators/with-expr
...
+
+- It contains a single field.
+- The struct doesn't specify an explicit layout.
+
+In addition, the compiler validates the <xref:System.Runtime.CompilerServices.InlineArrayAttribute?displayProperty=fullName> attribute:
+
+- The length must be > 0
+- The target type must be a struct.
+
+In most cases, and inline array can be accessed like an array, both to read and write values. In addition, you can use the [range](../operators/member-access-operators.md#range-operator-) and [index](../operators/member-access-operators.md#indexer-access) operators.
```
Comment (IEvangelist) [inline docs/csharp/language-reference/builtin-types/struct.md:101]:
```suggestion
In most cases, an inline array can be accessed like an array, both to read and write values. In addition, you can use the [range](../operators/member-access-operators.md#range-operator-) and [index](../operators/member-access-operators.md#indexer-access) operators.
```

File: docs/csharp/whats-new/csharp-12.md [hunk 2]
```diff
@@ -30,18 +31,49 @@ You can learn more about default parameters on lambda expressions in the article
 
 You can use the `using` alias directive to alias any type, not just named types. That means you can create semantic aliases for tuple types, array types, pointer types, or other unsafe types. For more information, see the [feature specification](~/_csharplang/proposals/using-alias-types.md)
 
+## Inline arrays
+
+Inline arrays are used by the runtime team and other library authors to improve performance in your application. Inline arrays enable a developer to create an array of fixed size in a `struct` type. A struct with an inline buffer should provide performance characteristics similar to an unsafe fixed size buffer. You likely won't declare your own inline arrays, but you'll use them transparently when they're exposed as <xref:System.Span%601?displayProperty=nameWithType> or <xref:System.ReadOnlySpan%601?displayProperty=nameWithType> objects from runtime APIs.
```
Comment (IEvangelist) [inline docs/csharp/whats-new/csharp-12.md:36]:
```suggestion
Inline arrays are used by the runtime team and other library authors to improve performance in your apps. Inline arrays enable a developer to create an array of fixed size in a `struct` type. A struct with an inline buffer should provide performance characteristics similar to an unsafe fixed size buffer. You likely won't declare your own inline arrays, but you'll use them transparently when they're exposed as <xref:System.Span%601?displayProperty=nameWithType> or <xref:System.ReadOnlySpan%601?displayProperty=nameWithType> objects from runtime APIs.
```

File: docs/csharp/whats-new/csharp-12.md [hunk 3]
```diff
@@ -30,18 +31,49 @@ You can learn more about default parameters on lambda expressions in the article
...
+var b = new Buffer();
+for(int i = 0; i < 10; i++)
+{
+    b[i] = i;
+}
+
+foreach(var i in b)
+{
+    Console.WriteLine(i);
+}
```
Comment (IEvangelist) [inline docs/csharp/whats-new/csharp-12.md:60]:
```suggestion
var buffer = new Buffer();
for (int i = 0; i < 10; i++)
{
    buffer[i] = i;
}

foreach (var i in buffer)
{
    Console.WriteLine(i);
}
```

**Review by IEvangelist — APPROVED** [review]
A few nits, otherwise let's :shipit:.

**Pull request merged**